File: apps/conclave-skip/Sources/Conclave/Core/Layout/GridLayout.swift
// Participant grid layout shared by SwiftUI and Skip/Compose.
// Keep the objective and tie-breaks in lockstep with the web grid layout.

import Foundation
#if canImport(CoreGraphics)
import CoreGraphics
#endif

struct GridLayoutOptions {
    var gap: CGFloat = 12
    var maxCols: Int = 7
    var maxTilesPerPage: Int = 49
    var targetAspect: CGFloat = 16.0 / 9.0
}

struct GridTilePosition {
    let index: Int
    let row: Int
    let col: CGFloat
    let x: CGFloat
    let y: CGFloat
    let width: CGFloat
    let height: CGFloat
}

struct GridLayoutResult {
    let cols: Int
    let rows: Int
    let tileWidth: CGFloat
    let tileHeight: CGFloat
    let lastRowCount: Int
    let pages: Int
    let perPage: Int
    let contentWidth: CGFloat
    let contentHeight: CGFloat
    let offsetX: CGFloat
    let offsetY: CGFloat
    let positions: [GridTilePosition]
}

private struct GridCandidate {
    let cols: Int
    let rows: Int
    let tileWidth: CGFloat
    let empty: Int
    let aspectDist: CGFloat
}

private struct GridTilePlacement {
    let contentWidth: CGFloat
    let contentHeight: CGFloat
    let offsetX: CGFloat
    let offsetY: CGFloat
    let positions: [GridTilePosition]
}

private func buildGridTilePositions(
    cols: Int,
    rows: Int,
    perPage: Int,
    tileWidth: CGFloat,
    tileHeight: CGFloat,
    gap: CGFloat,
    width: CGFloat,
    height: CGFloat
) -> GridTilePlacement {
    let contentWidth = max(0.0, CGFloat(cols) * tileWidth + CGFloat(max(0, cols - 1)) * gap)
    let contentHeight = max(0.0, CGFloat(rows) * tileHeight + CGFloat(max(0, rows - 1)) * gap)
    let offsetX = max(0.0, (width - contentWidth) / 2.0)
    let offsetY = max(0.0, (height - contentHeight) / 2.0)
    var positions: [GridTilePosition] = []

    var index = 0
    while index < perPage {
        let row = index / cols
        let rowStartIndex = row * cols
        let rowCount = min(cols, perPage - rowStartIndex)
        let rowWidth = CGFloat(rowCount) * tileWidth + CGFloat(max(0, rowCount - 1)) * gap
        let rowOffsetX = offsetX + max(0.0, (contentWidth - rowWidth) / 2.0)
        let col = index - rowStartIndex

        positions.append(GridTilePosition(
            index: index,
            row: row,
            col: CGFloat(col) + CGFloat(max(0, cols - rowCount)) / 2.0,
            x: rowOffsetX + CGFloat(col) * (tileWidth + gap),
            y: offsetY + CGFloat(row) * (tileHeight + gap),
            width: tileWidth,
            height: tileHeight
        ))
        index += 1
    }

    return GridTilePlacement(
        contentWidth: contentWidth,
        contentHeight: contentHeight,
        offsetX: offsetX,
        offsetY: offsetY,
        positions: positions
    )
}

/// Find the arrangement of `count` aspect-locked tiles that fits the largest
/// tile inside a `width` × `height` container. Objective: maximise the displayed
/// video area; ties break toward the grid whose bounding box best matches the
/// container aspect (so 2 people sit side-by-side in landscape), then fewer
/// empty cells.
func computeGridLayout(
    count: Int,
    width: CGFloat,
    height: CGFloat,
    options: GridLayoutOptions = GridLayoutOptions()
) -> GridLayoutResult {
    let gap = options.gap.isFinite && options.gap > 0.0 ? options.gap : 0.0
    let targetAspect = options.targetAspect.isFinite && options.targetAspect > 0.0
        ? options.targetAspect
        : 16.0 / 9.0
    let maxTilesPerPage = max(1, options.maxTilesPerPage)

    let total = max(1, count)
    let pages = Int(ceil(Double(total) / Double(maxTilesPerPage)))
    let perPage = min(total, maxTilesPerPage)

    // Degenerate container — return a single column so the caller still renders.
    if !width.isFinite || !height.isFinite || width <= 0 || height <= 0 {
        let placement = buildGridTilePositions(
            cols: 1, rows: perPage, perPage: perPage,
            tileWidth: 0.0, tileHeight: 0.0, gap: gap, width: 0.0, height: 0.0
        )
        return GridLayoutResult(
            cols: 1, rows: perPage, tileWidth: 0.0, tileHeight: 0.0,
            lastRowCount: 1, pages: pages, perPage: perPage,
            contentWidth: placement.contentWidth,
            contentHeight: placement.contentHeight,
            offsetX: placement.offsetX,
            offsetY: placement.offsetY,
            positions: placement.positions
        )
    }

    let colCap = min(perPage, max(1, options.maxCols))
    let containerAspect = width / height

    var best: GridCandidate? = nil

    var cols = 1
    while cols <= colCap {
        let rows = Int(ceil(Double(perPage) / Double(cols)))
        let cellW = (width - CGFloat(cols - 1) * gap) / CGFloat(cols)
        let cellH = (height - CGFloat(rows - 1) * gap) / CGFloat(rows)
        if cellW > 0 && cellH > 0 {
            // Largest aspect-locked tile that fits the cell (letterbox-fit).
            let tileWidth = min(cellW, cellH * targetAspect)
            let tileHeight = tileWidth / targetAspect
            let empty = cols * rows - perPage
            let boxW = CGFloat(cols) * tileWidth + CGFloat(cols - 1) * gap
            let boxH = CGFloat(rows) * tileHeight + CGFloat(rows - 1) * gap
            let aspectDist = abs(boxW / boxH - containerAspect)

            var better = best == nil
            if let b = best {
                if tileWidth > b.tileWidth + 0.5 {
                    better = true
                } else if abs(tileWidth - b.tileWidth) <= 0.5 {
                    if aspectDist < b.aspectDist - 0.01 {
                        better = true
                    } else if abs(aspectDist - b.aspectDist) <= 0.01 && empty < b.empty {
                        better = true
                    }
                }
            }
            if better {
                best = GridCandidate(cols: cols, rows: rows, tileWidth: tileWidth, empty: empty, aspectDist: aspectDist)
            }
        }
        cols += 1
    }

    let chosen = best ?? GridCandidate(cols: 1, rows: perPage, tileWidth: width, empty: 0, aspectDist: 0.0)
    let tileWidth = max(0.0, floor(chosen.tileWidth))
    let tileHeight = max(0.0, floor(tileWidth / targetAspect))
    let lastRowCount = perPage - (chosen.rows - 1) * chosen.cols
    let placement = buildGridTilePositions(
        cols: chosen.cols, rows: chosen.rows, perPage: perPage,
        tileWidth: tileWidth, tileHeight: tileHeight, gap: gap, width: width, height: height
    )

    return GridLayoutResult(
        cols: chosen.cols,
        rows: chosen.rows,
        tileWidth: tileWidth,
        tileHeight: tileHeight,
        lastRowCount: max(1, lastRowCount),
        pages: pages,
        perPage: perPage,
        contentWidth: placement.contentWidth,
        contentHeight: placement.contentHeight,
        offsetX: placement.offsetX,
        offsetY: placement.offsetY,
        positions: placement.positions
    )
}

// MARK: - Stage mode (Meet "Auto" selection above the grid packer)

enum StageMode {
    case tiled
    case spotlight
    case sideBySide
    case sidebar
}

struct StageModeInput {
    let count: Int
    let presenting: Bool
    let pinned: Bool
    let hasActiveVideoSpeaker: Bool
    var tiledThreshold: Int = 12
}

/// Meet "Auto" mode selection that runs ABOVE the grid packer. The packer is the
/// engine for `tiled` and for the people rail in `sidebar`/`sideBySide`.
func chooseStageMode(_ input: StageModeInput) -> StageMode {
    if input.pinned || input.count <= 2 || (input.presenting && !input.hasActiveVideoSpeaker) {
        return .spotlight
    }
    if input.presenting {
        return .sideBySide
    }
    if input.count <= input.tiledThreshold {
        return .tiled
    }
    return .sidebar
}

File: packages/meeting-core/src/grid-layout.ts
/**
 * Google-Meet-style participant grid arrangement — shared, pure, dependency-free.
 *
 * This is the client-side "tiled / auto" layout engine: given a visible tile
 * count, a measured container, and a target tile aspect ratio, it finds the
 * column/row arrangement that MAXIMISES tile size (the canonical optimal-grid
 * packing that Meet/Jitsi/Daily all use), then reports last-row centering and
 * paging so the renderer can lay it out exactly like Meet.
 *
 * The web client imports this directly; the Skip/SwiftUI app mirrors it in
 * Swift (GridLayout.swift) and verifies the same behavior in native tests.
 */

export interface GridLayoutOptions {
  /** Gap between tiles in px. Subtracted from the container before dividing. */
  gap?: number;
  /** Hard cap on columns (device class: ~6–7 desktop, 2–3 narrow phone). */
  maxCols?: number;
  /** Tiles shown per page before paging/overflow kicks in. */
  maxTilesPerPage?: number;
  /** Target tile aspect ratio = width / height (16:9 = 1.7778 landscape). */
  targetAspect?: number;
}

export interface GridTilePosition {
  /** Zero-based tile order on the current page. */
  index: number;
  /** Zero-based visual row. */
  row: number;
  /** Zero-based visual column within the logical grid. Partial rows may be fractional. */
  col: number;
  /** Left offset inside the measured layout area, in px. */
  x: number;
  /** Top offset inside the measured layout area, in px. */
  y: number;
  /** Tile width in px. */
  width: number;
  /** Tile height in px. */
  height: number;
}

export interface GridLayoutResult {
  /** Columns and rows for a full page. */
  cols: number;
  rows: number;
  /** Size of each tile in px (already aspect-constrained to targetAspect). */
  tileWidth: number;
  tileHeight: number;
  /** Tiles in the final (possibly partial) row — render centered. */
  lastRowCount: number;
  /** Number of pages when count exceeds maxTilesPerPage. */
  pages: number;
  /** Tiles laid out on a full page (<= maxTilesPerPage). */
  perPage: number;
  /** Width of the centered tile group, in px. */
  contentWidth: number;
  /** Height of the centered tile group, in px. */
  contentHeight: number;
  /** Left offset of the centered tile group inside the measured layout area. */
  offsetX: number;
  /** Top offset of the centered tile group inside the measured layout area. */
  offsetY: number;
  /** Exact tile positions for the current page, including centered partial rows. */
  positions: GridTilePosition[];
}

const DEFAULTS: Required<GridLayoutOptions> = {
  gap: 12,
  maxCols: 7,
  maxTilesPerPage: 49,
  targetAspect: 16 / 9,
};

const buildGridTilePositions = ({
  cols,
  rows,
  perPage,
  tileWidth,
  tileHeight,
  gap,
  width,
  height,
}: {
  cols: number;
  rows: number;
  perPage: number;
  tileWidth: number;
  tileHeight: number;
  gap: number;
  width: number;
  height: number;
}) => {
  const contentWidth = Math.max(0, cols * tileWidth + Math.max(0, cols - 1) * gap);
  const contentHeight = Math.max(0, rows * tileHeight + Math.max(0, rows - 1) * gap);
  const offsetX = Math.max(0, (width - contentWidth) / 2);
  const offsetY = Math.max(0, (height - contentHeight) / 2);
  const positions: GridTilePosition[] = [];

  for (let index = 0; index < perPage; index += 1) {
    const row = Math.floor(index / cols);
    const rowStartIndex = row * cols;
    const rowCount = Math.min(cols, perPage - rowStartIndex);
    const rowWidth = rowCount * tileWidth + Math.max(0, rowCount - 1) * gap;
    const rowOffsetX = offsetX + Math.max(0, (contentWidth - rowWidth) / 2);
    const col = index - rowStartIndex;

    positions.push({
      index,
      row,
      col: col + Math.max(0, cols - rowCount) / 2,
      x: rowOffsetX + col * (tileWidth + gap),
      y: offsetY + row * (tileHeight + gap),
      width: tileWidth,
      height: tileHeight,
    });
  }

  return {
    contentWidth,
    contentHeight,
    offsetX,
    offsetY,
    positions,
  };
};

/**
 * Find the arrangement of `count` aspect-locked tiles that fits the largest
 * tile inside a `width` × `height` container.
 *
 * The objective is "maximise the side of the largest tile that fits N copies"
 * (i.e. maximise the displayed-video area). For each candidate column count we
 * derive rows = ceil(N/cols), compute the largest tile that fits the cell while
 * respecting `targetAspect`, and keep the candidate whose tile is biggest;
 * ties break toward fewer empty cells (fuller grid).
 */
export function computeGridLayout(
  count: number,
  width: number,
  height: number,
  options: GridLayoutOptions = {},
): GridLayoutResult {
  const { gap, maxCols, maxTilesPerPage, targetAspect } = { ...DEFAULTS, ...options };

  const total = Math.max(1, Math.floor(count));
  const pages = Math.ceil(total / maxTilesPerPage);
  const perPage = Math.min(total, maxTilesPerPage);

  // Degenerate container — return a single column so the caller still renders.
  if (!Number.isFinite(width) || !Number.isFinite(height) || width <= 0 || height <= 0) {
    return {
      cols: 1,
      rows: perPage,
      tileWidth: 0,
      tileHeight: 0,
      lastRowCount: 1,
      pages,
      perPage,
      contentWidth: 0,
      contentHeight: 0,
      offsetX: 0,
      offsetY: 0,
      positions: buildGridTilePositions({
        cols: 1,
        rows: perPage,
        perPage,
        tileWidth: 0,
        tileHeight: 0,
        gap,
        width: 0,
        height: 0,
      }).positions,
    };
  }

  const colCap = Math.min(perPage, Math.max(1, Math.floor(maxCols)));
  const containerAspect = width / height;

  let best:
    | { cols: number; rows: number; tileWidth: number; empty: number; aspectDist: number }
    | null = null;

  for (let cols = 1; cols <= colCap; cols++) {
    const rows = Math.ceil(perPage / cols);
    const cellW = (width - (cols - 1) * gap) / cols;
    const cellH = (height - (rows - 1) * gap) / rows;
    if (cellW <= 0 || cellH <= 0) continue;

    // Largest aspect-locked tile that fits the cell (letterbox-fit).
    const tileWidth = Math.min(cellW, cellH * targetAspect);
    const tileHeight = tileWidth / targetAspect;
    const empty = cols * rows - perPage;
    // How well the grid's bounding box matches the container — this is the
    // tie-break that makes e.g. 2 people sit side-by-side in a landscape
    // container (both arrangements give the same tile size; the wider box wins).
    const boxW = cols * tileWidth + (cols - 1) * gap;
    const boxH = rows * tileHeight + (rows - 1) * gap;
    const aspectDist = Math.abs(boxW / boxH - containerAspect);

    let better = best === null;
    if (best !== null) {
      if (tileWidth > best.tileWidth + 0.5) better = true;
      else if (Math.abs(tileWidth - best.tileWidth) <= 0.5) {
        // Near-tie on tile size → best aspect match, then fuller grid.
        if (aspectDist < best.aspectDist - 0.01) better = true;
        else if (Math.abs(aspectDist - best.aspectDist) <= 0.01 && empty < best.empty)
          better = true;
      }
    }
    if (better) best = { cols, rows, tileWidth, empty, aspectDist };
  }

  // Fallback (shouldn't happen): one column.
  const chosen = best ?? { cols: 1, rows: perPage, tileWidth: width };

  const tileWidth = Math.max(0, Math.floor(chosen.tileWidth));
  const tileHeight = Math.max(0, Math.floor(tileWidth / targetAspect));
  const lastRowCount = perPage - (chosen.rows - 1) * chosen.cols;
  const placement = buildGridTilePositions({
    cols: chosen.cols,
    rows: chosen.rows,
    perPage,
    tileWidth,
    tileHeight,
    gap,
    width,
    height,
  });

  return {
    cols: chosen.cols,
    rows: chosen.rows,
    tileWidth,
    tileHeight,
    lastRowCount: Math.max(1, lastRowCount),
    pages,
    perPage,
    ...placement,
  };
}

export interface StageRailLayoutInput {
  /** Remote tiles competing for the stage rail, excluding the stage main tile. */
  candidateCount: number;
  /** Non-remote rail tiles that must remain visible, such as self-view/share. */
  fixedTileCount?: number;
  /** User/device tile budget for the whole stage layout. */
  maxTiles?: number;
  /** Measured rail viewport height in px. */
  railHeight: number;
  /** Compact rail tile height in px. */
  tileHeight?: number;
  /** Gap between rail tiles in px. */
  gap?: number;
}

export interface StageRailLayoutResult {
  /** Physical tile slots that fit in the measured rail viewport. */
  slotCount: number;
  /** Remote candidate tiles rendered as live video in the rail. */
  remoteCapacity: number;
  /** Whether a single overflow tile should reserve one visible rail slot. */
  overflowTile: boolean;
  /** Remote candidates hidden behind overflow. */
  hiddenCount: number;
  /** Fixed + remote + overflow tiles rendered in the rail. */
  renderedTileCount: number;
}

export function computeStageRailLayout(
  input: StageRailLayoutInput,
): StageRailLayoutResult {
  const candidateCount = Math.max(0, Math.floor(input.candidateCount));
  const fixedTileCount = Math.max(0, Math.floor(input.fixedTileCount ?? 0));
  const maxTiles = Number.isFinite(input.maxTiles ?? Number.POSITIVE_INFINITY)
    ? Math.max(0, Math.floor(input.maxTiles ?? 0))
    : Number.POSITIVE_INFINITY;
  const tileHeight = Math.max(1, Math.floor(input.tileHeight ?? 112));
  const gap = Math.max(0, Math.floor(input.gap ?? 12));
  const railHeight = Math.max(0, Math.floor(input.railHeight));

  const slotCount =
    railHeight > 0 ? Math.floor((railHeight + gap) / (tileHeight + gap)) : 0;
  const visibleSlots = Math.max(
    0,
    Math.min(slotCount - fixedTileCount, maxTiles - fixedTileCount),
  );
  const initialRemoteCapacity = Math.min(candidateCount, visibleSlots);
  const overflowTile =
    candidateCount > initialRemoteCapacity && visibleSlots > 0;
  const remoteCapacity = overflowTile
    ? Math.max(0, Math.min(candidateCount, visibleSlots - 1))
    : initialRemoteCapacity;
  const hiddenCount = Math.max(0, candidateCount - remoteCapacity);

  return {
    slotCount,
    remoteCapacity,
    overflowTile,
    hiddenCount,
    renderedTileCount:
      fixedTileCount + remoteCapacity + (overflowTile ? 1 : 0),
  };
}

export type StageMode = "tiled" | "spotlight" | "sideBySide" | "sidebar";

export interface StageModeInput {
  /** Visible participant count (incl. self). */
  count: number;
  /** Someone is screen-sharing / presenting. */
  presenting: boolean;
  /** A tile is explicitly pinned to the stage. */
  pinned: boolean;
  /** There is an active *video* speaker (not just the presentation). */
  hasActiveVideoSpeaker: boolean;
  /** Above this count (no presentation) Meet uses a sidebar instead of tiled. */
  tiledThreshold?: number;
}

/**
 * Meet "Auto" mode selection that runs ABOVE the grid packer. The packer is the
 * engine for the `tiled` mode and for the people rail in `sidebar`/`sideBySide`.
 */
export function chooseStageMode(input: StageModeInput): StageMode {
  const { count, presenting, pinned, hasActiveVideoSpeaker, tiledThreshold = 12 } = input;
  if (pinned || count <= 2 || (presenting && !hasActiveVideoSpeaker)) return "spotlight";
  if (presenting) return "sideBySide";
  if (count <= tiledThreshold) return "tiled";
  return "sidebar";
}
